internal/controller/validatorconfig_controller.go

```diff
@@ -93,12 +103,17 @@ func (r *ValidatorConfigReconciler) Reconcile(ctx context.Context, req ctrl.Requ
 		return ctrl.Result{}, removeFinalizer(ctx, r.Client, vc, CleanupFinalizer)
 	}
 
-	// ensure cleanup finalizer
-	if err := ensureFinalizer(ctx, r.Client, vc, CleanupFinalizer); err != nil {
-		r.Log.Error(err, "Error ensuring finalizer")
-		return ctrl.Result{}, err
-	}
-	r.Log.V(0).Info("Ensured ValidatorConfig finalizer")
+	// TODO: implement a proper patcher to avoid this hacky approach with retries & global vars
+	defer func() {
+		// Always update the ValidationConfig with a retry due to race condition with removeFinalizer.
+		for i := 0; i < constants.StatusUpdateRetries; i++ {
+			annotationErr := r.updateVc(ctx)
+			statusErr := r.updateVcStatus(ctx)
+			if annotationErr == nil && statusErr == nil {
+				break
+			}
```

Comment:
not sure what the nature of this race condition is, but would it make sense to add a small timeout between retries? or are there enough retries that the race condition is going to be resolved before we run out of retries?

Reply:
resolved in 4bc01cfb93a3970e4144cef2c7c03f2424494ff7